Source/Scene/BatchTable.js

```diff
@@ -0,0 +1,475 @@
+/*global define*/
+define([
+        '../Core/Cartesian2',
+        '../Core/Cartesian3',
+        '../Core/Cartesian4',
+        '../Core/combine',
+        '../Core/ComponentDatatype',
+        '../Core/defined',
+        '../Core/defineProperties',
+        '../Core/destroyObject',
+        '../Core/DeveloperError',
+        '../Core/PixelFormat',
+        '../Core/RuntimeError',
+        '../Renderer/ContextLimits',
+        '../Renderer/PixelDatatype',
+        '../Renderer/Sampler',
+        '../Renderer/Texture',
+        '../Renderer/TextureMagnificationFilter',
+        '../Renderer/TextureMinificationFilter'
+    ], function(
+        Cartesian2,
+        Cartesian3,
+        Cartesian4,
+        combine,
+        ComponentDatatype,
+        defined,
+        defineProperties,
+        destroyObject,
+        DeveloperError,
+        PixelFormat,
+        RuntimeError,
+        ContextLimits,
+        PixelDatatype,
+        Sampler,
+        Texture,
+        TextureMagnificationFilter,
+        TextureMinificationFilter) {
+    'use strict';
+
+    /**
+     * Creates a texture to look up per instance attributes for batched primitives. For example, store each primitive's pick color in the texture.
+     *
+     * @alias BatchTable
+     * @constructor
+     * @private
+     *
+     * @param {Object[]} attributes An array of objects describing a per instance attribute. Each object contains a datatype, components per attributes, whether it is normalized and a function name
+     *     to retrieve the value in the vertex shader.
+     * @param {Number} numberOfInstances The number of instances in a batch table.
+     *
+     * @example
+     * // create the batch table
+     * var attributes = [{
+     *     functionName : 'getShow()',
+     *     componentDatatype : ComponentDatatype.UNSIGNED_BYTE,
+     *     componentsPerAttribute : 1
+     * }, {
+     *     functionName : 'getPickColor',
+     *     componentDatatype : ComponentDatatype.UNSIGNED_BYTE,
+     *     componentsPerAttribute : 4,
+     *     normalize : true
+     * }];
+     * var batchTable = new BatchTable(attributes, 5);
+     *
+     * // when creating the draw commands, update the uniform map and the vertex shader
+     * vertexShaderSource = batchTable.getVertexShaderCallback()(vertexShaderSource);
+     * var shaderProgram = ShaderProgram.fromCache({
+     *    // ...
+     *    vertexShaderSource : vertexShaderSource,
+     * });
+     *
+     * drawCommand.shaderProgram = shaderProgram;
+     * drawCommand.uniformMap = batchTable.getUniformMapCallback()(uniformMap);
+     *
+     * // use the attribute function names in the shader to retrieve the instance values
+     * // ...
+     * attribute float batchId;
+     *
+     * void main() {
+     *     // ...
+     *     float show = getShow(batchId);
+     *     vec3 pickColor = getPickColor(batchId);
+     *     // ...
+     * }
+     */
+    function BatchTable(attributes, numberOfInstances) {
```

Comment:
Yes, it is `@private`.

I don't see why 3D Tiles wouldn't use this as a helper class. Unless you want to keep the option of not having to require VTF, then we could separate this out into helper functions.